Feature: Add python copyTo method to Tree and TreeNode

WhoBrokeTheBuild opened the pull request:
This adds the copyTo method to Tree and TreeNode in the python API copyTo allows you to copy a portion of a tree to another tree, optionally:
* Filtering unwanted nodes
* Copying data
* Copying tags
* Copying XNCIs
* Copying NCIs

Example:
```py
t1 = Tree('testsrc', -1, 'READONLY')
t2 = Tree('testdst', -1, 'EDIT')

def skip_devices(node):
    return (node.conglomerate_elt == 0)

t1.BRANCH.copyTo(t2.SECTION, node_filter=skip_devices, copy_xnci=False)

t1.close()
t2.write()
t2.close()
```

WhoBrokeTheBuild committed:
Feature: Add python copyTo method to Tree and TreeNode

This adds the copyTo method to Tree and TreeNode in the python API
copyTo allows you to copy a portion of a tree to another tree, optionally:
* Filtering unwanted nodes
* Copying data
* Copying tags
* Copying XNCIs
* Copying NCIs

Example:
```
t1 = Tree('testsrc', -1, 'READONLY')
t2 = Tree('testdst', -1, 'EDIT')

def skip_devices(node):
    return (node.conglomerate_elt == 0)

t1.BRANCH.copyTo(t2.SECTION, node_filter=skip_devices, copy_xnci=False)

t1.close()
t2.write()
t2.close()
```

mwinkel-dev reviewed (commented):
Just making sure that I understand how devices are handled . . . 

1. I gather that if the `node_filter` removes the head node of a "device", that the entire device is skipped (as per the example given in the feature's description).
2. However, if devices are copied, the latter steps of the procedure apply a sanity check to make sure that all of the associated conglomerate_elt nodes have been copied and are in the proper order.

Is my understanding correct?

WhoBrokeTheBuild commented:
Yes, the `node_filter` is unable to filter out part of a conglomerate, and yes the nodes are sorted very carefully so that they get the correct conglomerate_elt.

mwinkel-dev reviewed (dismissed):
The code is easy to understand and matches the feature description.   So, approving this PR.

heidthecamp commented on python/MDSplus/tree.py:
```diff
@@ -1182,6 +1182,10 @@ def tdiData(self, arg, **kwargs):
         kwargs['tree'] = self.tree
         return _dat.TdiData(arg, **kwargs)
 
+    def copyTo(src, dst, **kwargs):
```
Why are there two definitions for copyTo?

WhoBrokeTheBuild replied:
Good question, the one in `Tree` is a helper function for people who want to copy an entire tree, e.g. `t1.copyTo(t2)` or `t1.copyTo(t2.SOMEWHERE)`

WhoBrokeTheBuild committed:
Remove python 3.6 f-strings, replace mdsExceptions with _exc

heidthecamp commented on python/MDSplus/tree.py:
```diff
@@ -3102,6 +3106,201 @@ def updateSegmentLimits(self, start, end, idx):
+        for src_node in src_nodes:
+            dst_path = src_node.fullpath.replace(root_src, root_dst)
+
+            try:
```
Why are we not logging the exceptions?

WhoBrokeTheBuild replied:
Not logging, but the whole function will fail

WhoBrokeTheBuild committed:
Fix Typos

WhoBrokeTheBuild committed:
Rename src to self and add an alias

gtrevisan commented on python/MDSplus/tree.py:
```diff
@@ -1182,6 +1182,10 @@ def tdiData(self, arg, **kwargs):
         kwargs['tree'] = self.tree
         return _dat.TdiData(arg, **kwargs)
 
+    def copyTo(self, src, dst, **kwargs):
```
shouldn't you remove `src` from here?
I'd imagine our first (and only) non-kw arg, that is, the destination, would get assigned to `src` rather than `dst`...
is the alias going to solve it?

santorofer commented on python/MDSplus/tree.py:
```diff
@@ -3102,6 +3106,204 @@ def updateSegmentLimits(self, start, end, idx):
 
+    def copyTo(self, dst, node_filter=None, copy_tags=True, copy_data=True, copy_nci=True, copy_xnci=True):
+        """Recursively copy a portion of one tree to another.
+        @param src: The source TreeNode to copy nodes from
```
Maybe this line in the comments (i.e @param src: ...) should be removed and just add a comment below?

WhoBrokeTheBuild committed:
Fix Tree.copyTo extra argument

WhoBrokeTheBuild committed:
Update docstring for TreeNode.copyTo

WhoBrokeTheBuild committed:
Further sanitize tag names

santorofer reviewed (dismissed):
Looks very good now.

WhoBrokeTheBuild committed:
Update TreeNode.copyTo to update tree paths in data

Remove decompile/compile for node data, as this lost floating point precision
Add _update_tree_paths which recursively finds tree nodes in data structures and rewrites them to point at the destination tree

mwinkel-dev reviewed (dismissed):
The `_update_tree_paths()` function looks fine.   As do all previous commits.

WhoBrokeTheBuild committed:
Add missing getDefault() to _update_tree_paths

WhoBrokeTheBuild committed:
Add workaround for weird Tree/NID scenario
